### PR TITLE
[NOD-1147] Implement address exchange

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: protocol/receiveaddresses/receiveaddresses.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	peer *peerpkg.Peer, addressManager *addrmgr.AddrManager) (routeClosed bool, err error) {
+
+	subnetworkID, err := peer.SubnetworkID()
+	if err != nil {
+		panic(err)
+	}
+
+	needAddresses := addressManager.NeedMoreAddresses()
+
+	msgGetAddr := wire.NewMsgGetAddr(addressManager.NeedMoreAddresses(), false, subnetworkID)
```
Comment (svarogg) [inline protocol/receiveaddresses/receiveaddresses.go:26]: You just called `addressManager.NeedMoreAddresses()` a couple of lines above, and stored it in `needAddresses`

File: protocol/sendaddresses/sendaddresses.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	"time"
+)
+
+const timeout = 30 * time.Second
+
+// SendAddresses sends address to a peer that requests it.
+func SendAddresses(incomingRoute *router.Route, outgoingRoute *router.Route,
+	addressManager *addrmgr.AddrManager) (routeClosed bool, err error) {
+
+	message, isOpen, err := incomingRoute.DequeueWithTimeout(timeout)
```
Comment (svarogg) [inline protocol/sendaddresses/sendaddresses.go:17]: Remove timeout

File: protocol/sendaddresses/sendaddresses.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return false, err
+	}
+	if !isOpen {
+		return true, nil
+	}
+
+	msgGetAddr := message.(*wire.MsgGetAddr)
+	if !msgGetAddr.NeedAddresses {
```
Comment (svarogg) [inline protocol/sendaddresses/sendaddresses.go:26]: As agreed, the whole concept of needAddresses in the message is redundant

File: protocol/sendaddresses/sendaddresses.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		return addresses
+	}
+
+	shuffleAddresses := make([]*wire.NetAddress, addressCount)
+	copy(shuffleAddresses, addresses)
+
+	// Shuffle the address list.
+	for i := 0; i < wire.MaxAddrPerMsg; i++ {
+		j := i + rand.Intn(addressCount-i)
+		shuffleAddresses[i], shuffleAddresses[j] = shuffleAddresses[j], shuffleAddresses[i]
```
Comment (svarogg) [inline protocol/sendaddresses/sendaddresses.go:58]: Better to use rand.Shuffle (it doesn't help MUCH, but it's a bit shorter)

File: protocol/protocol.go [hunk 5]
```diff
@@ -136,3 +151,28 @@ func addFlow(name string, router *routerpkg.Router, messageTypes []string, stopp
 		}
 	})
 }
+
+func addOneTimeFlow(name string, router *routerpkg.Router, messageTypes []string, stopped *uint32,
```
[the remark this answers is not in the document]
Comment (someone235) [reply protocol/protocol.go:155]: No, addOneTimeFlow is async and you don't care when it's finished. It doesn't apply to handshake

File: protocol/protocol.go [hunk 6]
```diff
@@ -136,3 +151,28 @@ func addFlow(name string, router *routerpkg.Router, messageTypes []string, stopp
...
+	route, err := router.AddIncomingRoute(messageTypes)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		err := router.RemoveRoute(messageTypes)
+		if err != nil {
+			panic(err)
+		}
+	}()
```
Comment (stasatdaglabs) [inline protocol/protocol.go:167]: It appears that you remove the route immediately, as the rest of the function is async.

File: protocol/receiveaddresses/receiveaddresses.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+// ReceiveAddresses asks a peer for more addresses if needed.
+func ReceiveAddresses(incomingRoute *router.Route, outgoingRoute *router.Route,
+	peer *peerpkg.Peer, addressManager *addrmgr.AddrManager) (routeClosed bool, err error) {
+
+	subnetworkID, err := peer.SubnetworkID()
+	if err != nil {
+		panic(err)
+	}
+
+	msgGetAddr := wire.NewMsgGetAddr(false, subnetworkID)
```
Comment (stasatdaglabs) [inline protocol/receiveaddresses/receiveaddresses.go:24]: Please rename everything related to GetAddr to GetAddresses and everything related to Addr to Addresses.

File: protocol/sendaddresses/sendaddresses.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package sendaddresses
+
+import (
+	"github.com/kaspanet/kaspad/addrmgr"
+	"github.com/kaspanet/kaspad/netadapter/router"
+	"github.com/kaspanet/kaspad/wire"
+	"math/rand"
+)
+
+// SendAddresses sends address to a peer that requests it.
```
Comment (stasatdaglabs) [inline protocol/sendaddresses/sendaddresses.go:10]: Typo: address -> addresses

File: protocol/sendaddresses/sendaddresses.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+	msgGetAddr := message.(*wire.MsgGetAddr)
+	addresses := addressManager.AddressCache(msgGetAddr.IncludeAllSubnetworks, msgGetAddr.SubnetworkID)
+	msgAddr := wire.NewMsgAddr(msgGetAddr.IncludeAllSubnetworks, msgGetAddr.SubnetworkID)
+	err = msgAddr.AddAddresses(shuffleAddresses(addresses)...)
+	if err != nil {
+		panic(err)
+	}
+
+	isOpen = outgoingRoute.Enqueue(msgAddr)
```
Comment (stasatdaglabs) [inline protocol/sendaddresses/sendaddresses.go:27]: Please use EnqueueWithTimeout here.

File: protocol/receiveaddresses/receiveaddresses.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+	message, isOpen, err := incomingRoute.DequeueWithTimeout(timeout)
+	if err != nil {
+		return false, err
+	}
+	if !isOpen {
+		return true, nil
+	}
+
+	msgAddr := message.(*wire.MsgAddr)
```
Comment (stasatdaglabs) [inline protocol/receiveaddresses/receiveaddresses.go:45]: Please rename MsgAddr to MsgAddresses.

File: protocol/receiveaddresses/receiveaddresses.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	if !msgAddr.SubnetworkID.IsEqual(config.ActiveConfig().SubnetworkID) && msgAddr.SubnetworkID != nil {
+		return false, protocolerrors.Errorf(false, "only full nodes and %s subnetwork IDs "+
+			"are allowed in [%s] command, but got subnetwork ID %s",
+			config.ActiveConfig().SubnetworkID, msgAddr.Command(), msgAddr.SubnetworkID)
+	}
+
+	// TODO(libp2p) Consider adding to peer known addresses set
+	// TODO(libp2p) Replace with real peer IP
+	fakeSrcAddr := new(wire.NetAddress)
```
Comment (stasatdaglabs) [inline protocol/receiveaddresses/receiveaddresses.go:62]: Please rename this to fakeSourceAddress.

File: protocol/sendaddresses/sendaddresses.go [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+// SendAddresses sends addresses to a peer that requests it.
+func SendAddresses(incomingRoute *router.Route, outgoingRoute *router.Route,
+	addressManager *addrmgr.AddrManager) (routeClosed bool, err error) {
+
+	message, isOpen := incomingRoute.Dequeue()
+	if !isOpen {
+		return true, nil
+	}
+
+	msgGetAddr := message.(*wire.MsgGetAddresses)
```
Comment (stasatdaglabs) [inline protocol/sendaddresses/sendaddresses.go:20]: Please rename msgGetAddr to msgGetAddresses.

File: protocol/sendaddresses/sendaddresses.go [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	addresses := addressManager.AddressCache(msgGetAddr.IncludeAllSubnetworks, msgGetAddr.SubnetworkID)
+	msgAddr := wire.NewMsgAddr(msgGetAddr.IncludeAllSubnetworks, msgGetAddr.SubnetworkID)
```
Comment (stasatdaglabs) [inline protocol/sendaddresses/sendaddresses.go:22]: Please rename NewMsgAddr to NewMsgAddresses.

File: wire/message_test.go [hunk 14]
```diff
@@ -54,7 +54,7 @@ func TestMessage(t *testing.T) {
 	msgVersion := NewMsgVersion(me, idMe, &daghash.ZeroHash, nil)
 
 	msgVerack := NewMsgVerAck()
-	msgGetAddr := NewMsgGetAddr(false, nil)
+	msgGetAddr := NewMsgGetAddresses(false, nil)
```
Comment (stasatdaglabs) [inline wire/message_test.go:57]: Please rename msgGetAddr to msgGetAddresses.

File: wire/msggetaddr_test.go [hunk 15]
```diff
@@ -13,15 +13,15 @@ import (
 	"github.com/kaspanet/kaspad/util/subnetworkid"
 )
 
-// TestGetAddr tests the MsgGetAddr API.
+// TestGetAddr tests the MsgGetAddresses API.
 func TestGetAddr(t *testing.T) {
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:17]: Please rename TestGetAddr to TestGetAddresses.

File: wire/msggetaddr_test.go [hunk 16]
```diff
@@ -36,18 +36,18 @@ func TestGetAddr(t *testing.T) {
 	}
 }
 
-// TestGetAddrWire tests the MsgGetAddr wire encode and decode for various
+// TestGetAddrWire tests the MsgGetAddresses wire encode and decode for various
 // protocol versions.
 func TestGetAddrWire(t *testing.T) {
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:41]: Please rename TestGetAddrWire to TestGetAddressesWire.

File: wire/msggetaddr_test.go [hunk 17]
```diff
@@ -36,18 +36,18 @@ func TestGetAddr(t *testing.T) {
...
 	// With all subnetworks
-	msgGetAddr := NewMsgGetAddr(false, nil)
+	msgGetAddr := NewMsgGetAddresses(false, nil)
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:43]: Please rename this variable.

File: wire/msggetaddr_test.go [hunk 18]
```diff
@@ -36,18 +36,18 @@ func TestGetAddr(t *testing.T) {
...
 	msgGetAddrEncoded := []byte{
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:44]: Please rename this variable.

File: wire/msggetaddr_test.go [hunk 19]
```diff
@@ -36,18 +36,18 @@ func TestGetAddr(t *testing.T) {
...
 		0x00, // All subnetworks
 		0x01, // Get full nodes
 	}
 
 	// With specific subnetwork
-	msgGetAddrSubnet := NewMsgGetAddr(false, subnetworkid.SubnetworkIDNative)
+	msgGetAddrSubnet := NewMsgGetAddresses(false, subnetworkid.SubnetworkIDNative)
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:50]: Please rename this variable.

File: wire/msggetaddr_test.go [hunk 20]
```diff
@@ -36,18 +36,18 @@ func TestGetAddr(t *testing.T) {
...
 	msgGetAddrSubnetEncoded := []byte{
```
Comment (stasatdaglabs) [inline wire/msggetaddr_test.go:51]: Please rename this variable.

File: protocol/sendaddresses/sendaddresses.go [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+		return true, nil
+	}
+	return false, nil
+}
+
+// shuffleAddresses randomizes the given addresses sent if there are more than the maximum allowed in one message.
+func shuffleAddresses(addresses []*wire.NetAddress) []*wire.NetAddress {
+	addressCount := len(addresses)
+
+	if addressCount < wire.MaxAddrPerMsg {
```
Comment (stasatdaglabs) [inline protocol/sendaddresses/sendaddresses.go:43]: Please rename MaxAddrPerMsg to MaxAddressesPerMessage.

File: wire/msgaddr.go [hunk 22]
```diff
@@ -158,21 +158,21 @@ func (msg *MsgAddr) KaspaEncode(w io.Writer, pver uint32) error {
...
-// Message interface. See MsgAddr for details.
-func NewMsgAddr(includeAllSubnetworks bool, subnetworkID *subnetworkid.SubnetworkID) *MsgAddr {
-	return &MsgAddr{
+// NewMsgAddresses returns a new kaspa addr message that conforms to the
+// Message interface. See MsgAddresses for details.
+func NewMsgAddresses(includeAllSubnetworks bool, subnetworkID *subnetworkid.SubnetworkID) *MsgAddresses {
+	return &MsgAddresses{
 		IncludeAllSubnetworks: includeAllSubnetworks,
 		SubnetworkID:          subnetworkID,
 		AddrList:              make([]*NetAddress, 0, MaxAddrPerMsg),
```
Comment (stasatdaglabs) [inline wire/msgaddr.go:178]: Please rename AddrList too Addresses.

File: wire/msgaddr_test.go [hunk 23]
```diff
@@ -17,15 +17,15 @@ import (
 	"github.com/kaspanet/kaspad/util/subnetworkid"
 )
 
-// TestAddr tests the MsgAddr API.
+// TestAddr tests the MsgAddresses API.
 func TestAddr(t *testing.T) {
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:21]: Please rename TestAddr to TestAddresses.

File: wire/msgaddr_test.go [hunk 24]
```diff
@@ -17,15 +17,15 @@ import (
 	"github.com/kaspanet/kaspad/util/subnetworkid"
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:17]: Please correct the name of this file.

File: wire/msgaddr.go [hunk 25]
```diff
@@ -12,10 +12,10 @@ import (
 )
```
Comment (stasatdaglabs) [inline wire/msgaddr.go:12]: Please correct the name of this file.

File: wire/msgaddr_test.go [hunk 26]
```diff
@@ -75,7 +75,7 @@ func TestAddr(t *testing.T) {
 	}
 }
 
-// TestAddrWire tests the MsgAddr wire encode and decode for various numbers
+// TestAddrWire tests the MsgAddresses wire encode and decode for various numbers
 // of addresses and protocol versions.
 func TestAddrWire(t *testing.T) {
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:80]: Please rename TestAddrWire to TestAddressesWire.

File: wire/msgaddr_test.go [hunk 27]
```diff
@@ -93,15 +93,15 @@ func TestAddrWire(t *testing.T) {
 	}
 
 	// Empty address message.
-	noAddr := NewMsgAddr(false, nil)
+	noAddr := NewMsgAddresses(false, nil)
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:96]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 28]
```diff
@@ -93,15 +93,15 @@ func TestAddrWire(t *testing.T) {
...
 	noAddrEncoded := []byte{
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:97]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 29]
```diff
@@ -93,15 +93,15 @@ func TestAddrWire(t *testing.T) {
...
 		0x00, // All subnetworks
 		0x01, // Is full node
 		0x00, // Varint for number of addresses
 	}
 
 	// Address message with multiple addresses.
-	multiAddr := NewMsgAddr(true, nil)
+	multiAddr := NewMsgAddresses(true, nil)
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:104]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 30]
```diff
@@ -119,7 +119,7 @@ func TestAddrWire(t *testing.T) {
 	}
 
 	// Address message with multiple addresses and subnetworkID.
-	multiAddrSubnet := NewMsgAddr(false, subnetworkid.SubnetworkIDNative)
+	multiAddrSubnet := NewMsgAddresses(false, subnetworkid.SubnetworkIDNative)
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:122]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 31]
```diff
@@ -203,7 +203,7 @@ func TestAddrWire(t *testing.T) {
 }
 
 // TestAddrWireErrors performs negative tests against wire encode and decode
-// of MsgAddr to confirm error paths work correctly.
+// of MsgAddresses to confirm error paths work correctly.
 func TestAddrWireErrors(t *testing.T) {
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:207]: Please rename TestAddrWireErrors to TestAddressesWireErrors.

File: wire/msgaddr_test.go [hunk 32]
```diff
@@ -223,7 +223,7 @@ func TestAddrWireErrors(t *testing.T) {
 	}
 
 	// Address message with multiple addresses.
-	baseAddr := NewMsgAddr(false, nil)
+	baseAddr := NewMsgAddresses(false, nil)
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:226]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 33]
```diff
@@ -223,7 +223,7 @@ func TestAddrWireErrors(t *testing.T) {
...
 	baseAddr.AddAddresses(na, na2)
 	baseAddrEncoded := []byte{
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:228]: Please rename this variable.

File: wire/msgaddr_test.go [hunk 34]
```diff
@@ -242,7 +242,7 @@ func TestAddrWireErrors(t *testing.T) {
 
 	// Message that forces an error by having more than the max allowed
 	// addresses.
-	maxAddr := NewMsgAddr(false, nil)
+	maxAddr := NewMsgAddresses(false, nil)
```
Comment (stasatdaglabs) [inline wire/msgaddr_test.go:245]: Please rename this variable.

**Pull request merged**